nilearn/image/image.py

```diff
@@ -502,6 +502,61 @@ def swap_img_hemispheres(img):
 
     # create swapped nifti object
     out_img = nibabel.Nifti1Image(img.get_data()[::-1], img.get_affine(),
-        header=img.get_header())
+                                  header=img.get_header())
 
     return out_img
+
+
+def _index_img(imgs, index):
+    """Helper function for index_img and iter_img."""
+    return nibabel.Nifti1Image(imgs.get_data()[:, :, :, index],
+                               imgs.get_affine(),
+                               header=imgs.get_header())
+
+
+def index_img(imgs, index):
+    """Indexes into a 4D Niimg-like object in the fourth dimension.
+
+    Common use cases include extracting a 3D image out of `imgs` or
+    creating a 4D image whose data is a subset of `imgs` data.
+
+    Parameters
+    ----------
+    imgs: 4D Niimg-like object
+          See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+
+    index: Any type compatible with numpy array indexing
+           Used for indexing the 4D data array in the fourth dimension.
+
+    Returns
+    -------
+    output: nibabel.Nifti1Image
+
+    """
+    imgs = check_niimgs(imgs)
+    return _index_img(imgs, index)
+
+
+class _Checked4DImageIterator(object):
+    """Helper class for iter_img, use iter_img directly and not this class."""
+    def __init__(self, imgs):
+        self.imgs = check_niimgs(imgs)
```

Comment:
Same remark here: the data gets loaded when the iterator is created. Not only does it get loaded, but it gets concatenated to a 4D image. That means that it is using way more memory than it should be.

We should try to find a way around this.